### PR TITLE
Add resource monitor for testplan

### PR DESCRIPTION
**Pull request by yuxuan-ms**
Add resource monitor that will export a JSON file includes:

* test and Testplan internals (events) start and stop times (seconds)
* resource utilisation per host

To run Testplan with Resource Monitor just run your Testplan script with the --resource-monitor command line argument.

### CODE REVIEW

File: testplan/common/utils/monitor/__init__.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+    def _record_event(self, event_uuid, event):
+        if event_uuid not in self.events_data:
+            self.events_data[event_uuid] = []
+        self.events_data[event_uuid].append({
+            'event': event,
+            'time': time.time(),
+        })
+
+    def started(self, event_type, metadata=None):
```
Comment (Pyifan) [inline testplan/common/utils/monitor/__init__.py]: Consider rename started -> start_event? otherwise it is confusing when resource_monitor has a start() method that does totally different thing.

same for stopped

File: testplan/testing/multitest/base.py [hunk 2]
```diff
@@ -333,6 +335,10 @@ def run_tests(self):
         ctx = [(self.test_context[idx][0], self.test_context[idx][1][:])
                for idx in range(len(self.test_context))]
 
+        if self.cfg.resource_monitor:
```
Comment (Pyifan) [inline testplan/testing/multitest/base.py]: Can we handle other tests like GTest too?

Or if we want finer granularity e.g suite/case level, would current design be easy to adapt?
Maybe do we need to record parent-child relationship via json?

File: testplan/common/entity/base.py [hunk 3]
```diff
@@ -458,6 +458,8 @@ def make_runpath_dirs(self):
         Creates runpath related directories.
         """
         self._runpath = self.generate_runpath()
+        # Update self.cfg.runpath if use the default runpath
+        self.cfg.runpath = self._runpath
```
Comment (JohnChiotis) [inline testplan/common/entity/base.py]: Configuration input should not be altered

File: testplan/common/utils/monitor/__init__.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+from testplan.common.utils.path import makedirs
+from .collectors import get_collector
+
+
+class EventMonitor(object):
+    """
+    Event monitor will record event start and stop times(seconds)
+    """
+    def __init__(self, parent=None):
+        self.logger = logging.getLogger('Monitor')
```
Comment (JohnChiotis) [inline testplan/common/utils/monitor/__init__.py]: I think in this implementation we may have to use the global TESTPLAN_LOGGER object that repo offers and just register a 'monitor' file handler?
from testplan.logger import TESTPLAN_LOGGER

File: testplan/common/utils/monitor/__init__.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        :type log_path: ``str``
+        :param log_level: Level of logging to save. Currently either ``logging.DEBUG`` if "debug" given or
+                          ``logging.INFO`` if anything else is given.
+        :type log_level: ``str``
+
+        :return: ``None``
+        :rtype: ``NoneType``
+        """
+        self.logger = logging.getLogger(str(uuid.uuid4()))
+        if log_level == 'debug':
```
Comment (JohnChiotis) [inline testplan/common/utils/monitor/__init__.py]: Will have to change based on the global logger.

File: testplan/common/utils/monitor/__init__.py (hunk 1 above)
Comment (JohnChiotis) [reply testplan/common/utils/monitor/__init__.py]: event_started()

File: testplan/common/utils/monitor/__init__.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+            elif isinstance(metadata, dict):
+                _metadata = metadata.copy()
+            else:
+                raise TypeError('Event metadata must be a dictionary - not {}'.format(type(metadata)))
+            _metadata['type'] = event_type
+            self.events_metadata[event_uuid] = _metadata
+            self._record_event(event_uuid, 'start')
+        return event_uuid
+
+    def stopped(self, event_uuid, metadata=None):
```
Comment (JohnChiotis) [inline testplan/common/utils/monitor/__init__.py]: event_stopped()

File: testplan/common/utils/monitor/__init__.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        Start the thread of collect the metrics.
+
+        :return: ``None``
+        :rtype: ``NoneType``
+        """
+        self._poll = threading.Thread(target=self._monitoring)
+        self._poll.daemon = True
+        self._poll_start = True
+        self._poll.start()
+        self.logger.info('Resource polling started')
```
Comment (JohnChiotis) [inline testplan/common/utils/monitor/__init__.py]: dot . end of sentence

File: testplan/common/utils/monitor/collectors.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    def __init__(self, director):
+        super(LinuxCollector, self).__init__(director)
+        try:
+            cpu_info = subprocess.check_output(['lscpu'])
+            self.metadata['extra']['cpu_info'] = cpu_info.decode()
+        except Exception:
+            pass
+
+
+class LinuxContainerCollector(LinuxCollector):
```
Comment (JohnChiotis) [inline testplan/common/utils/monitor/collectors.py]: Is this relevant to OSS world? Are the file paths standard? should they be configurable?

File: testplan/common/utils/monitor/collectors.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                except (IOError, AttributeError, ValueError):
+                    pass
+
+
+class WindowsCollector(Collector):
+    def __init__(self, director):
+        super(WindowsCollector, self).__init__(director)
+
+
+def get_collector(director=os.getcwd()):
```
Comment (JohnChiotis) [inline testplan/common/utils/monitor/collectors.py]: we use an internal utility instead of os.getcwd()
https://github.com/Morgan-Stanley/testplan/blob/master/testplan/common/utils/path.py#L36

File: testplan/common/utils/monitor/collectors.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    if platform.system() == 'Windows':
+        return WindowsCollector(director)
+    elif platform.system() == 'Linux':
+        try:
+            with open('/proc/1/cgroup') as cgroup_file:
```
Comment (JohnChiotis) [inline testplan/common/utils/monitor/collectors.py]: Same for this path

File: testplan/testing/multitest/base.py (hunk 2 above)
Comment (JohnChiotis) [reply testplan/testing/multitest/base.py]: Should we have the monitor logic in the Test() base class of all MultiTest, GTest etc?

File: testplan/common/utils/monitor/__init__.py [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+        """
+        self.logger = logging.getLogger(str(uuid.uuid4()))
+        if log_level == 'debug':
+            log_level = logging.DEBUG
+        else:
+            log_level = logging.INFO
+        if log_path:
+            directory = log_path.rsplit(os.sep, 1)[0]
+            self.log_directory = directory
+            makedirs(directory)
```
Comment (JohnChiotis) [inline testplan/common/utils/monitor/__init__.py]: Check for race conditions

File: testplan/common/utils/monitor/__init__.py [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+            file_handler = logging.FileHandler(filename=log_path)
+        else:
+            file_handler = logging.NullHandler()  # pylint: disable=bad-option-value
+        formatter = logging.Formatter('%(asctime)s-Monitor-%(levelname)-8s:  %(message)s')
+        file_handler.setFormatter(formatter)
+        file_handler.setLevel(log_level)
+        self.logger.setLevel(log_level)
```
Comment (JohnChiotis) [inline testplan/common/utils/monitor/__init__.py]: if self logger is global, that needs to go

**Comment by ryan-collingham** [comment]
This has been open for a long time - what's happening with the resource monitor?

**Comment by yuxuan-ms** [comment]
> This has been open for a long time - what's happening with the resource monitor?

Let's close the pull request. After complete the report UI, I will create a new pull request.
